### PR TITLE
[SPARK-7478][SQL] Added SQLContext.getOrCreate

### PR DESCRIPTION
**Pull request by tdas**
Having a SQLContext singleton would make it easier for applications to use a lazily instantiated single shared instance of SQLContext when needed. It would avoid problems like 
1. In REPL/notebook environment, rerunning the line {{val sqlContext = new SQLContext}} multiple times created different contexts while overriding the reference to previous context, leading to issues like registered temp tables going missing.
2. In Streaming, creating SQLContext directly leads to serialization/deserialization issues when attempting to recover from DStream checkpoints. See [SPARK-6770]. Also to get around this problem I had to suggest creating a singleton instance - https://github.com/apache/spark/blob/master/examples/src/main/scala/org/apache/spark/examples/streaming/SqlNetworkWordCount.scala

This can be solved by {{SQLContext.getOrCreate}} which get or creates a new singleton instance of SQLContext using either a given SparkContext or a given SparkConf.

@rxin @marmbrus

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 1]
```diff
@@ -1272,6 +1273,56 @@ class SQLContext(@transient val sparkContext: SparkContext)
     }
   }
 
+  SQLContext.setLastInstantiatedContext(self)
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: @rxin  Design question. Should contexts created directly (that is, not through `getOrCreate`) also be automatically considered for the singleton?

**Comment by marmbrus** [comment]
What about `HiveContext`?

**Comment by tdas** [comment]
@marmbrus Good you raised the question. @rxin and I had half-a-discussion offline about it. Correct me if I am wrong but HiveContext is a superset of the SQLContext in terms of functionality. So the last HiveContext created should also be accessible through this interface. Since the SQLContext constructor register itself with singleton, the HiveContext will also do the same. This part made sense to me. 

Going beyond that if you think there should also be a HiveContext.getOrCreate, which would return the last instantiated HiveContext (not SQLContext), I can add that too in this PR.

**Comment by marmbrus** [comment]
Yeah the first part seems reasonable.  But I think you also need a way to force the getOrCreate command to create a hive context.

**Comment by tdas** [comment]
test this please.

**Comment by tdas** [comment]
So `HiveContext.getOrCreate` sounds good?

**Comment by tdas** [comment]
@marmbrus Can you help me debug this?

**Comment by scwf** [comment]
HiveContext use derby as the metastore database defaultly which only support single instance for one metastore path.

I think you need config the ```javax.jdo.option.ConnectionURL``` before creating the HiveContext
you can change getorcreate like this:
```
  def getOrCreate(sparkContext: SparkContext, config: Map[String, String] = Map.empty): HiveContext   = {
    INSTANTIATION_LOCK.synchronized {
      if (lastInstantiatedContext.get() == null) {
        new HiveContext(sparkContext) {
          override def configure(): Map[String, String] = config
        }
      }
    }
    lastInstantiatedContext.get()
  }
```

then when you create hivecontext in HiveContextSuite, you can configure it like this
```
    val hiveContext = HiveContext.getOrCreate(testSparkContext, HiveContext.newTemporaryConfiguration)

```

**Comment by tdas** [comment]
@marmbrus Could you take a look?

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 2]
```diff
@@ -1270,9 +1271,59 @@ class SQLContext(@transient val sparkContext: SparkContext)
...
+
+  private val INSTANTIATION_LOCK = new Object()
+
+  /**
+   * Reference to the last created SQLContext.
+   */
+  @transient private val lastInstantiatedContext = new AtomicReference[SQLContext]()
 
+  /**
+   * Get the singleton SQLContext if it exists or create a new one using the given configuration.
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: using the given `SparkContext`.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 3]
```diff
@@ -1270,9 +1271,59 @@ class SQLContext(@transient val sparkContext: SparkContext)
...
+  def getOrCreate(sparkContext: SparkContext): SQLContext = {
+    INSTANTIATION_LOCK.synchronized {
+      if (lastInstantiatedContext.get() == null) {
+        new SQLContext(sparkContext)
+      }
+    }
+    lastInstantiatedContext.get()
+  }
+
+  private[sql] def getLastInstantiatedContext(): Option[SQLContext] = {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Why have a private method that is not used anywhere?

**Comment by marmbrus** [comment]
Minor comments, otherwise LGTM.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 3 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Aah, was using it for tests but changed tests in the middle.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 2 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Good catch.

**Comment by tdas** [comment]
I merging this! Thanks @marmbrus for reviewing.

**Comment by jelez** [comment]
Can you guys point me what is the proper way to use getOrCreate with HiveContext?

**Comment by mwws** [comment]
@jelez you can create an HiveContextSingleton to workaround it. Refer to example "SqlNetWorkWordCount"

@tdas Why you removed HiveContext.getOrCreate? I can't find obvious reasons from the conversation.

**Pull request closed without merging**